42.md

```diff
@@ -18,6 +18,42 @@ A relay may want to require clients to authenticate to access restricted resourc
 
 ## Definitions
 
+### New field to identify protected events
+
+Clients that want to force an authentication in any event can add a `requires_auth_by` with array of authorized pubkeys to the base event object. 
+
+Relays MUST authenticate connections with at least one of the pubkeys listed in `requires_auth_by` to return events.
+
+```jsonc
+{
+  "id": <32-bytes lowercase hex-encoded sha256 of the serialized event data>,
+  "pubkey": <32-bytes lowercase hex-encoded public key of the event creator>,
+  "requires_auth_by" [<32-bytes lowercase hex-encoded public key>, /* ... */ ],
+  "created_at": <unix timestamp in seconds>,
+  "kind": <integer between 0 and 65535>,
+  "tags": [
+    [<arbitrary string>...],
+    // ...
+  ],
+  "content": <arbitrary string>,
+  "sig": <64-bytes lowercase hex of the signature of the sha256 hash of the serialized event data, which is the same as the "id" field>
+}
+```
+
+NIP-42 supporting relays must then include the `requires_auth_by` field as the 7th position in the serialization of the event. This guarantees that relays and clients that do not support this NIP cannot validate protected events and will never be able to receive them, avoiding leaking the event structure to a public relay. 
+
+```js
+[
+  0,
+  <pubkey, as a lowercase hex string>,
+  <created_at, as a number>,
+  <kind, as a number>,
+  <tags, as an array of arrays of non-null strings>,
+  <content, as a string>,
+  <requires_auth_by, as an array of non-null strings> 
+]
+```
+
```

Comment:
```suggestion
### Private Event Kind Range

Events in the kind range `40000-49999` are considered always private. They MUST have atleast one `A` tag. Each of these tags is set to a `pubkey` allowed to request the event after authenticating using the `AUTH` flow defined on below section.

These events should be treated by `relays` as parameterized replaceable events. When a new NIP chooses a kind number in this range for an event, it should define if `clients` will consider that the event will:

1) Act as a non-replaceable event, by being referenced by its `id` and by using a random `d` tag value OR;
2) Act as a replaceable event, by using a `d` tag set to an empty string OR;
3) Act as a parameterized replaceable event, by using a non-fixed `d` tag value OR;
4) Act as an ephemeral event, by using an `expiration` tag set to a value soon to expire, by being referenced by its `id` and by using a random `d` tag value.

```

Reply:
@vitorpamplona what do you think? Using an event kind in a specific kind range to signal expected relay behavior does have the benefit of not requiring relay builders to check every new NIP in look for how they should treat each newly added event kind.

Removed the serialization in favor of simply condering those relays that advertise NIP-42 support on "supported_nips" as able to enforce event access control.

Reply:
@vitorpamplona please take a closer look at the following paragraph "These events should be treated by `relays` as parameterized replaceable events. When a new NIP chooses a kind number in this range for an event, it should define if `clients` will consider that the event will: 1) **Act as a non-replaceable** event [...]"

A parameterized replaceable event is flexible enough to be treated by `clients` as non-replaceable, replaceable or ephemeral. While relays just treat them as parameterized replaceable ones.

Reply:
nah, that is too confusing. Better not have a range at all then

Reply:
I mean, there is no need to create 4 ranges (regular, replaceable, param repl. and ephemeral) but now requiring AUTH.

Reply:
Why not just add the AUTH requirement option to any event? There is no need to recreate event kinds that already exist just to make them authed.

Reply:
> ideally relays should know in advance if user should AUTH just by looking at the REQ filter"

I don't think this is true. Once the relay gets the event that match a filter, it can remove the ones that have the `A` if the connection is not authed. Or if you are using an SQL database, it's super easy to know if the user is auth, include As, if it is not, then remove them.

Reply:
> Once the relay gets the event that match a filter, it can remove the ones that have the A if the connection is not authed

This is ambiguous behavior. The client won't be able to know the cause of the missing events. Was it because the event didn't exist on the relay or the client had no authorization to fetch them?
Also, will the specific relay implementation do as you said and reply with some of the events or instead simply send none? The client can't be sure so can't code effectively a retry strategy.

This kind of situation was the reason we had to add the `CLOSED` message to NIP-42.

Reply:
I mean, today if a client isn't authenticated and asks for `{ kinds: [1, 4] }`, the relay should send `CLOSED` instead of just the `kind:1` events.

And that being the case, for relay implementation simplicity, if it receives a REQ filter without `kinds`, it may choose to (and frankly should) send `CLOSED` right away. Better than getting events from DB just to later find out it includes a private event, using resources for no reason, just to later discard events and send `CLOSED`.

Reply:
> This is ambiguous behavior. The client won't be able to know the cause of the missing events.

This already happens all the time. It's nothing new. Half of relays send missing events even though we have received them before with the same filter. 

> Also, will the specific relay implementation do as you said and reply with some of the events or instead simply send none?

Yep. Simply send an EOSE if you don't have any. It already happens today with DM relays that filter out GiftWraps to the pubkey alone. 

> I mean, today if a client isn't authenticated and asks for { kinds: [1, 4] }, the relay should send CLOSED instead of just the kind:1 events.

That's bad. It should send the 1s and the EOSE when done. We mix in a lot of filters to go around subscription and number of filter limits per relay. Almost all of our filters include private and public events. It's virtually impossible to have something like Amethyst without merging several data needs in the same subscription.

Reply:
FYI, I just looked at the guidance for CLOSED and it doesn't include "relay should send CLOSED instead of just the kind:1 events". And I think that should never be the case. If the relay is refusing to fulfill the subscription, better find something better than not accepting one of the filters.

Reply:
There were 3 PRs and other ideas trying to fix ambiguous AUTH flow until @fiatjaf came up with the new message.

My understanding and why i liked the solution is that `CLOSED` refers to a subscription as a whole and the client retries a sub with exactly the same filter after authenticating. That's because the message includes the sub id and the reason is just a brief code plus free text (doesn't explain the reason in detail for a machine to read - e.g.: you will know you need to auth but not which events triggered it): `["CLOSED", "sub_ID", "reason"]`

> Almost all of our filters include private and public events

The idea is if you haven't already authenticated, you would receive a `CLOSED`, authenticate then retry it with a `REQ` with same filter.

If you get some events and then EOSE it means you don't need to retry and many clients will certainly close the subscription at this moment.

Reply:
Opened #1046 PR editing the example to make it more clear.

Reply:
Opened #1049 PR to better explain the motivation